src/api/index.ts

```diff
@@ -0,0 +1,184 @@
+      method: 'POST',
+      body: JSON.stringify({
+        data: {
+          type: 'release',
+          attributes: params.attributes,
+          relationships: {
+            product: {
+              data: {
+                type: 'product',
+                id: params.productID
+              }
+            },
+            constraints: {
+              data: []
+            }
+          }
+        }
+      })
+    })
+
+    if (!response.ok) {
+      throw new KeygenAPIError(response)
+    }
+
+    const {data = undefined, errors = undefined} = (await response.json()) as {
+      data?: {id: string}
+      errors?: unknown[]
+    }
+
+    if (errors?.length || !data) {
+      throw new KeygenAPIError(response)
+    }
+
+    return data
+  }
+
+  async releasePublish(
+    params: InferType<typeof releasePublishSchema>
+  ): Promise<void> {
+    const response = await this._makeAPIRequest({
+      resourcePath: `/releases/${params.releaseID}/actions/publish`,
+      method: 'POST'
+    })
+
+    if (!response.ok) {
+      throw new KeygenAPIError(response)
+    }
+
+    const {errors = undefined} = (await response.json()) as {
+      errors?: unknown[]
+    }
+
+    if (errors?.length) {
+      throw new KeygenAPIError(response)
+    }
+
+    return
+  }
+
+  async artifactCreate(
+    params: InferType<typeof artifactCreateSchema>
+  ): Promise<{
+    id: string
+    links: {
+      redirect: string
+    }
+  }> {
+    const response = await this._makeAPIRequest({
+      resourcePath: '/artifacts',
+      method: 'POST',
+      body: JSON.stringify({
+        data: {
+          type: 'artifact',
+          attributes: params.attributes,
+          relationships: {
+            release: {
+              data: {
+                type: 'release',
+                id: params.releaseID
+              }
+            }
+          }
+        }
+      }),
+      redirect: 'manual' // TODO: explain
+    })
+
+    if (response.status !== 307) {
+      throw new KeygenAPIError(response)
+    }
+
+    const {data = undefined, errors = undefined} = (await response.json()) as {
+      data?: {
+        id: string
+        links: {
+          redirect: string
+        }
+      }
+      errors?: unknown[]
+    }
+
+    if (errors?.length || !data) {
+      throw new KeygenAPIError(response)
+    }
+
+    return data
+  }
+
+  async artifactFileUpload(
+    signedS3UploadUrl: string,
+    file: Buffer
+  ): Promise<void> {
+    // NOTE:
+    // This endpoint expects the whole file to be loaded in memory.
+    // As a result, the upload operation is at the mercy of available Runner memory.
+    //
+    // We would've made the upload read the file as a stream (using fs.createReadStream()). However, S3 does not support this.
+    // We would've also considered supporting S3's Multipart upload (when a file is detected to be larger than a given threshold). Howeber, Keygen does not provide utilities for getting pre-signed part upload urls.
```

Comment:
What does S3 not support about it? Can't you send a streaming HTTP PUT regardless of what the API supports - this seems like a major limitation?

Reply:
My bad, S3 supports streaming uploads just fine. Implemented in https://github.com/unhack/keygen-action/pull/11/commits/00251801f5a0eeb019bb7f1effbeac09d297e943